org.eclipse.lsp4j.websockets/src/main/java/org/eclipse/lsp4j/launch/websockets/LSPWebSocketServer.java

```diff
@@ -0,0 +1,133 @@
+import javax.websocket.OnOpen;
+import javax.websocket.Session;
+
+import org.eclipse.lsp4j.jsonrpc.Endpoint;
+import org.eclipse.lsp4j.jsonrpc.JsonRpcException;
+import org.eclipse.lsp4j.jsonrpc.MessageConsumer;
+import org.eclipse.lsp4j.jsonrpc.MessageIssueException;
+import org.eclipse.lsp4j.jsonrpc.RemoteEndpoint;
+import org.eclipse.lsp4j.jsonrpc.json.JsonRpcMethod;
+import org.eclipse.lsp4j.jsonrpc.json.MessageConstants;
+import org.eclipse.lsp4j.jsonrpc.json.MessageJsonHandler;
+import org.eclipse.lsp4j.jsonrpc.messages.Message;
+import org.eclipse.lsp4j.jsonrpc.messages.NotificationMessage;
+import org.eclipse.lsp4j.jsonrpc.messages.RequestMessage;
+import org.eclipse.lsp4j.jsonrpc.messages.ResponseMessage;
+import org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint;
+import org.eclipse.lsp4j.jsonrpc.services.ServiceEndpoints;
+import org.eclipse.lsp4j.services.LanguageClient;
+import org.eclipse.lsp4j.services.LanguageClientAware;
+import org.eclipse.lsp4j.services.LanguageServer;
+
+public class LSPWebSocketServer<T extends LanguageServer> {
+
+        private static final Logger LOG = Logger.getLogger(LSPWebSocketServer.class.getName());
+
+        private final Map<Session,Endpoint> endpoints = new LinkedHashMap<>();
+
+	private final MessageJsonHandler jsonHandler;
+	private final Supplier<T> newServer;
+	
+	LSPWebSocketServer(Supplier<T> newServer, Class<T> serverClass) {
+		this.newServer = newServer;
+		
+		Map<String, JsonRpcMethod> supportedMethods = new LinkedHashMap<>();
+		supportedMethods.putAll(ServiceEndpoints.getSupportedMethods(LanguageClient.class));
+		supportedMethods.putAll(ServiceEndpoints.getSupportedMethods(serverClass));
+		
+		this.jsonHandler = new MessageJsonHandler(supportedMethods);
+	}
+	
+	@OnOpen
+    public void onOpen(Session session) throws IOException{
+        System.out.println("Open Connection " + session + " ...");
+
+        LanguageServer server = newServer.get();
+		Endpoint local = new GenericEndpoint(Collections.singleton(server));
+		endpoints.put(session, local);
+		
+		if (server instanceof LanguageClientAware) {
+			MessageConsumer out = new MessageConsumer() {
+				@Override
+				public void consume(Message arg0) throws MessageIssueException, JsonRpcException {
+				    LOG.log(Level.INFO, "sending message: " + arg0);
+					try {
+						session.getBasicRemote().sendText(arg0.toString());
+					} catch (IOException e) {
+						throw new Error(e);
+					}
+				}
+			};
+
+			RemoteEndpoint remote = new RemoteEndpoint(out, local);
+			((LanguageClientAware)server).connect(ServiceEndpoints.toServiceObject(remote, LanguageClient.class));
+		}
+    }
+
+    @OnClose
+    public void onClose(Session session) throws IOException{
+        System.out.println("Close Connection " + session + " ...");
+        endpoints.remove(session);
+        session.close();
+    }
+
+    @OnMessage
+    public String onMessage(String message, Session session) throws InterruptedException, ExecutionException{
```

Comment:
This method lacks error handling for messages (see StreamMessageProducer) and validation (ReflectiveMessageValidator).